File: Data Structures & Algorithms/number-of-students-unable-to-eat-lunch/submission-1.cpp
/*
    Approach: Using Two Deque
    Time complexity: O(n^2)
    Space complexity: O(n)
*/

#include <deque>
using namespace std;

class Solution {
public:
    int countStudents(vector<int>& students, vector<int>& sandwiches) {
        deque<int> sandwich_queue(sandwiches.begin(), sandwiches.end()); 
        deque<int> student_queue(students.begin(), students.end()); 
        int distance_of_unhappy_student = -1;

        while (student_queue.size() and distance_of_unhappy_student != student_queue.size()) {
            if (student_queue[0] == sandwich_queue[0]) {
                student_queue.pop_front();
                sandwich_queue.pop_front();
                distance_of_unhappy_student = -1;
            }
            else {
                int front = student_queue.front();  
                student_queue.pop_front();          
                student_queue.push_back(front);
                distance_of_unhappy_student++;
            }
        }

        return student_queue.size();
    }
};

File: Data Structures & Algorithms/number-of-students-unable-to-eat-lunch/submission-1.py
# Took hint

from collections import deque


class Solution:
    def countStudents(self, students: List[int], sandwiches: List[int]) -> int:
        student_count = [0,0]

        for student in students:
            student_count[student]+=1

        for sandwich in sandwiches:
            if student_count[sandwich] == 0:
                break

            student_count[sandwich]-=1

        return sum(student_count)